File: Day 1 - Arrays I/75. Sort Colors/Solution.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    void sortColors(vector<int> &nums) {
        int low = 0, mid = 0, high = nums.size() - 1;
        while (mid <= high) {
            if (nums[mid] == 0) {
                swap(nums[low], nums[mid]);
                low++;
                mid++;
            } else if (nums[mid] == 1) {
                mid++;
            } else {
                swap(nums[mid], nums[high]);
                high--;
            }
        }
    }

    void SortColorsDNF(vector<int> &nums) {
        int low = 0, mid = 0, high = nums.size() - 1;

        for (int i = 0; i < nums.size(); ++i) {
            if (nums[mid] == 0) {
                swap(nums[mid], nums[low]);
                low++;
                mid++;
            } else if (nums[mid] == 1) {
                mid++;
            } else {
                swap(nums[mid], nums[high]);
                high--;
            }
        }
    }
};

int main() {
    Solution solution;
    vector<int> nums = {2, 0};
//    solution.sortColors(nums);
    solution.SortColorsDNF(nums);
    for (int i = 0; i < nums.size(); i++) {
        cout << nums[i] << " ";
    }
    return 0;
}

File: Day 1 - Arrays I/75. Sort Colors/Solution.py
class Solution:
    def SortColors(self, nums: [int]):
        rez = {0: 0, 1: 0, 2: 0}

        for i in nums:
            rez[i] += 1

        k = 0
        for i in range(len(nums)):
            if rez[k] != 0:
                nums[i] = k
                rez[k] -= 1
            else:
                k += 1
                nums[i] = k
                rez[k] -= 1

        print(nums)


if __name__ == '__main__':
    solution = Solution()

    solution.SortColors([0, 2])
